### PR TITLE
KAFKA-13769: Explicitly route FK join results to correct partitions

### PR DESCRIPTION
**Pull request by Gerrrr**
Prior to this commit FK response sink routed FK results to
SubscriptionResolverJoinProcessorSupplier using the primary key.

There are cases, where this behaviour is incorrect. For example,
if KTable key serde differs from the data source serde which might
happen without a key changing operation.

Instead of determining the resolver partition by serializing the PK
this patch includes target partition in SubscriptionWrapper and
SubscriptionResponseWrapper payloads. Default FK response-sink
partitioner extracts the correct partition from the value and
routes the message accordingly.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java [hunk 1]
```diff
@@ -23,12 +23,13 @@
...
 
 public class SubscriptionWrapper<K> {
-    static final byte CURRENT_VERSION = 0;
+    static final byte CURRENT_VERSION = 1;
 
     private final long[] hash;
     private final Instruction instruction;
     private final byte version;
     private final K primaryKey;
+    private final Integer primaryPartition;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java:34]: Maybe add a comment: `// added in v1 -- null for v0

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java [hunk 2]
```diff
@@ -79,16 +79,27 @@ public void setIfUnset(final SerdeGetter getter) {
             );
 
             final ByteBuffer buf;
+            int dataLength = 2 + primaryKeySerializedData.length;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: Can we unify the `dataLength` computation into a single place? It's a little bit scattered at hard to follow. Also with different versions, I like a more explicit pattern using switch:
```
switch (version) {
case 0: encodeV1(); break;
case 1: encodeV2(); break;
default: throw...
}
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 3]
```diff
@@ -1197,7 +1197,7 @@ boolean sendingOldValueEnabled() {
 
         final StreamPartitioner<K, SubscriptionResponseWrapper<VO>> foreignResponseSinkPartitioner =
             tableJoinedInternal.partitioner() == null
-                ? null
+                ? (topic, key, subscriptionResponseWrapper, numPartitions) -> subscriptionResponseWrapper.getPrimaryPartition()
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1200]: Just wanted to confirm my understanding -- we originally returned `null` what implies that the producer will call it's configured partitioner to compute the target partition number. Using the new custom partitioner implies that if we get an old response that has no "primary partition" encoded, we still return `null` via `getPrimaryPartition()` and thus it's fully backward compatible?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ public void setIfUnset(final SerdeGetter getter) {
 
         @Override
         public byte[] serialize(final String topic, final SubscriptionResponseWrapper<V> data) {
-            //{1-bit-isHashNull}{7-bits-version}{Optional-16-byte-Hash}{n-bytes serialized data}
+            //{1-bit-isHashNull}{7-bits-version}{4-bytes-primaryPartition}{Optional-16-byte-Hash}{n-bytes serialized data}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: I think we cannot just write the new format blindly. The receiving task may be executed on a different instance that might not have been ungraded yet during a rolling upgrade. -- We can only start to use the new format after all instances are upgraded, otherwise older instances would crash (or the upgrade can not be done in rolling bounce fashion what seen not acceptable).

Not sure right now how we can handle this case though: The coarse grained solution would be to use a config and let users do two rolling bounces (requires a KIP). Or we try to piggy-back on the rebalance metadata version, but I am not sure how easy it would be to do, given that we setup the serdes at a totally different layer and get access to the rebalance metadata version only during runtime.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: Yeah, this is indeed a problem. It's the same issue I documented here: https://issues.apache.org/jira/browse/KAFKA-10336

I was also thinking that the rebalance metadata version could give us a mechanism to coordinate this kind of thing across a rolling bounce. I documented the approach in that ticket, but never got a chance to work on it.

If we don't want to bite off a general solution to it, two rolling bounces might be the best compromise, as annoying as they are.

Alternatively, since we actually do check the version and throw an exception if we encounter it, it is "safe" (in that there's no data corruption) to just go ahead and let the exceptions happen during a rolling bounce. Not the nicest thing to do, though.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 3 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java:1200]: Yes, this is correct. For completeness, here is how I reasoned about it. Without this patch the partitioner is `null`, so `RecordCollectorImpl` sets the partition to `null` here:

https://github.com/apache/kafka/blob/ce883892270a02a72e91afbdb1fabdd50d7da474/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L136

With this patch, the partitioner is always not `null`, so `RecordCollector` uses it to find desired partition here:

https://github.com/apache/kafka/blob/ce883892270a02a72e91afbdb1fabdd50d7da474/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L130

In the case when the new partitioner returns `null` for the v0 data, we get to the same state as without this patch - the following call to another version of `RecordCollectorImpl#send` has `partition` set to `null`:

https://github.com/apache/kafka/blob/ce883892270a02a72e91afbdb1fabdd50d7da474/streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java#L139

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java (hunk 1 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java:34]: Added in fcfe805cd1847cb9f441535a30c6a8b87c27e8e1

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java (hunk 2 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: Done in a59a4f6fe2a0d61d15656742f1d42862918f80a6

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java (hunk 4 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: What will happen when an old processor receives a v1 record and throws an exception? Will it throw that record on the floor or just blocks processing until the upgrade?

As John suggested in [KAFKA-10336](https://issues.apache.org/jira/browse/KAFKA-10336), we can pass `UPGRADE_FROM` into the serde so that it produces v0 records if the option is set. I will also add `3.3` as a valid choice for that option. WDYT?

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: > What will happen when an old processor receives a v1 record and throws an exception? Will it throw that record on the floor or just blocks processing until the upgrade?

The thread with crash, and we will eventually rebalance. So there is no data-loss, but for at-least-once duplicate processing. It's very bad user experience if they upgrade one instance and suddenly all other might crash...

Re-using `UPGRAD_FROM` should work. The config is already there, so maybe it's ok if we just add a never "accepted value" without doing a KIP? We need to cover this case in the upgrade docs though!

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java [hunk 5]
```diff
@@ -92,6 +103,32 @@ public void setIfUnset(final SerdeGetter getter) {
...
+            final int dataLength = 1 + hashLength + serializedDataLength + primaryPartitionLength;
+
+            final ByteBuffer buf = ByteBuffer.allocate(dataLength);
+
+            if (originalHash != null) {
+                buf.put(data.getVersion());
+            } else {
+                buf.put((byte) (data.getVersion() | (byte) 0x80));
+            }
+            buf.putInt(data.getPrimaryPartition());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: Can we move this to the end of the byte array and thus re-use `serializeV0` to write the prefix?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java (hunk 4 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: Added `UPGRADE_FROM` in [d9f6c52](https://github.com/apache/kafka/pull/11945/commits/d9f6c52e2fa3a7ebd46efdade63a0c392acc8d0c). There, I added all versions from 2.7 to 3.2 as `UPGRADE_FROM` options. If serde detects the upgrade, it uses v0 format to serialize the data. Deserializer does not need any changes as there is no difference between v0 and upgrade-from-v0 formats.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java (hunk 5 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: I am not sure if this is a good idea. As an advantage, we will remove duplicated parts in `serialize`. However, there are 2 downsides:
* We use v0 path in 2 situations - when it is *real* v0 and when we upgrade from v0 (i.e. a data record can actually have a primary partition and v1 as part of its state). This means that we can't just put `data.getVersion()` into the buffer, but have to hardcode it to `0`. Making this logic generic enough to handle v0, upgrade-from, and v1 will be cumbersome.
* Moving a fixed-size field to the end of the byte array will complicate deserialization. Consider

https://github.com/apache/kafka/blob/d9f6c52e2fa3a7ebd46efdade63a0c392acc8d0c/streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java#L234-L242

Right now, the logic for finding PK size is straightforward. It can be summarized as "take it from here and until the end of the buffer". If we move primary partition to the end, we'll have to remove another 4 bytes iff the version is greater than 0.

I'd say that code duplication is lesser evil in this case. WDYT? I am happy to move primary partition to the end if you don't agree.

File: docs/streams/upgrade-guide.html [hunk 6]
```diff
@@ -34,9 +34,9 @@ <h1>Upgrade Guide and API Changes</h1>
     </div>
 
     <p>
-        Upgrading from any older version to {{fullDotVersion}} is possible: if upgrading from 2.3 or below, you will need to do two rolling bounces, where during the first rolling bounce phase you set the config <code>upgrade.from="older version"</code>
-        (possible values are <code>"0.10.0" - "2.3"</code>) and during the second you remove it. This is required to safely upgrade to the new cooperative rebalancing protocol of the embedded consumer. Note that you will remain using the old eager
-        rebalancing protocol if you skip or delay the second rolling bounce, but you can safely switch over to cooperative at any time once the entire group is on 2.4+ by removing the config value and bouncing. For more details please refer to
+        Upgrading from any older version to {{fullDotVersion}} is possible: if upgrading from 3.3 or below, you will need to do two rolling bounces, where during the first rolling bounce phase you set the config <code>upgrade.from="older version"</code>
```
Comment (mjsax) [inline docs/streams/upgrade-guide.html]: > upgrading from 3.3

Should be `3.2` -- the fix should go into 3.3, right?

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java [hunk 7]
```diff
@@ -636,7 +684,10 @@
         "Accepted values are \"" + UPGRADE_FROM_0100 + "\", \"" + UPGRADE_FROM_0101 + "\", \"" +
         UPGRADE_FROM_0102 + "\", \"" + UPGRADE_FROM_0110 + "\", \"" + UPGRADE_FROM_10 + "\", \"" +
         UPGRADE_FROM_11 + "\", \"" + UPGRADE_FROM_20 + "\", \"" + UPGRADE_FROM_21 + "\", \"" +
-        UPGRADE_FROM_22 + "\", \"" + UPGRADE_FROM_23 + "\" (for upgrading from the corresponding old version).";
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:639]: > When upgrading from 2.4 to a newer version it is not required to specify this config.

Seems this needs an update?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java [hunk 8]
```diff
@@ -62,6 +71,42 @@ public void setIfUnset(final SerdeGetter getter) {
...
         }
 
+        @Override
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            this.upgradeFromV0 = upgradeFromV0(configs);
+        }
+
+        private static boolean upgradeFromV0(final Map<String, ?> configs) {
+            final Object upgradeFrom = configs.get(StreamsConfig.UPGRADE_FROM_CONFIG);
+            if (!(upgradeFrom instanceof String)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: We don't need this check. `StreamsConfig` ensures that it will be a `String`. Cf `StreamConfig`:
```
.define(UPGRADE_FROM_CONFIG,
        ConfigDef.Type.STRING,
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java [hunk 9]
```diff
@@ -60,6 +63,42 @@ public void setIfUnset(final SerdeGetter getter) {
...
         }
 
+        @Override
+        public void configure(final Map<String, ?> configs, final boolean isKey) {
+            this.upgradeFromV0 = upgradeFromV0(configs);
+        }
+
+        private static boolean upgradeFromV0(final Map<String, ?> configs) {
+            final Object upgradeFrom = configs.get(StreamsConfig.UPGRADE_FROM_CONFIG);
+            if (!(upgradeFrom instanceof String)) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: As above.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java (hunk 5 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: >  This means that we can't just put data.getVersion() into the buffer, but have to hardcode it to 0

Yes, but this seems ok? Ie, we would have a `serializeV0Internal(final short version)` that is called as (something like):
```
public byte[] serializeV0() {
  return serializeV0Internal(0).getBytes();
}

public byte[] serializeV1() {
  final ByteStream out = serializeV0Internal(0);
  
  // add v1 stuff to `out`
  
  return out.getBytes();
}
```

> Making this logic generic enough to handle v0, upgrade-from, and v1 will be cumbersome.

Why?

I understand that putting variable length at the end could be simpler. I guess I am just used to how it's usually done if Kafka. And personally, if I read code it's easier to have a mental model of the different versions in the bytes, if older versions are always a prefix of newer versions?

Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: Fixed in 13e772e9d075d4c7f609eacf9855c0c0afb9ad2b.

File: docs/streams/upgrade-guide.html (hunk 6 above)
Comment (Gerrrr) [reply docs/streams/upgrade-guide.html]: Right! Fixed in cd8ba1848cefca1c89b4011e9d0716b8534f6782.

File: streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java (hunk 7 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:639]: Good catch! Fixed in cd8ba1848cefca1c89b4011e9d0716b8534f6782.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java (hunk 8 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: Replaced with a not-null check in 8ebbaaf38ee79642ccb322694de7556f260f93ca.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java (hunk 9 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: Replaced with a not-null check in https://github.com/apache/kafka/commit/8ebbaaf38ee79642ccb322694de7556f260f93ca.

**Comment by Gerrrr** [comment]
@mjsax Can you please take another look?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java [hunk 10]
```diff
@@ -21,22 +21,30 @@
 import java.util.Arrays;
 
 public class SubscriptionResponseWrapper<FV> {
-    final static byte CURRENT_VERSION = 0x00;
+    final static byte CURRENT_VERSION = 0x01;
+    // 0x00
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java]: nit:
```
// version 0x00 fields:
```

Same below:
```
// version 0x01 fields:
```

Or just change to `1` instead of `0x01` (compare Subscription.java -- might be nice to align both).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java [hunk 11]
```diff
@@ -21,22 +21,30 @@
...
     private final long[] originalValueHash;
     private final FV foreignValue;
     private final byte version;
+    // 0x01
+    private final Integer primaryPartition;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java:30]: While we need this field in this class, so we actually need to serialize it? It seems it must not go over the wire, does it?

Thus, for the subscription response, we don't need the dump the version? -- We only need it on the right hand side to write the response into the topic, but we can do this in-memory only?

If this is correct, we also don't need to update the corresponding Serde.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java [hunk 12]
```diff
@@ -23,12 +23,15 @@
 
 
 public class SubscriptionWrapper<K> {
-    static final byte CURRENT_VERSION = 0;
+    static final byte CURRENT_VERSION = 1;
 
+    // v0
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java]: nit:
```
// v0 fields:
```

Same for `v1` below

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java [hunk 13]
```diff
@@ -141,17 +204,29 @@ public SubscriptionWrapper<K> deserialize(final String ignored, final byte[] dat
...
+            final byte[] primaryKeyRaw = new byte[primaryKeyLength];
+            buf.get(primaryKeyRaw, 0, primaryKeyLength);
 
             if (primaryKeySerializationPseudoTopic == null) {
                 primaryKeySerializationPseudoTopic = primaryKeySerializationPseudoTopicSupplier.get();
             }
 
             final K primaryKey = primaryKeyDeserializer.deserialize(primaryKeySerializationPseudoTopic,
-                                                                    primaryKeyRaw);
+                primaryKeyRaw);
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: nit: indention

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java [hunk 14]
```diff
@@ -62,36 +71,96 @@ public void setIfUnset(final SerdeGetter getter) {
...
+                throw new UnsupportedVersionException("Unsupported SubscriptionWrapper version " + data.getVersion());
+            }
+        }
+
+        private ByteBuffer serializeCommon(
+            final String topic,
+            final SubscriptionResponseWrapper<V> data,
+            final byte version,
+            final int extraLength) {
             final byte[] serializedData = data.getForeignValue() == null ? null : serializer.serialize(topic, data.getForeignValue());
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: nit (formatting -- hard to see where method signature end and code starts):
```
    final int extraLength
) {
    final byte[] serializedData = data.getForeignValue() == null ? null : serializer.serialize(topic, data.getForeignValue());
```

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java [hunk 15]
```diff
@@ -132,18 +201,31 @@ public SubscriptionResponseWrapper<V> deserialize(final String topic, final byte
             }
 
             final V value;
-            if (data.length - lengthSum > 0) {
+            final int valueLength;
+            if (version > 0) {
+                valueLength = data.length - lengthSum - Integer.BYTES;
+
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapperSerde.java]: nit: remove empty line

**Review by mjsax — COMMENTED** [review]
Overall LGTM -- just wondering if we need to change the subscription-response wire format?

Also (can also be a follow up PR): we should add a system test for the upgrade to make sure it really works) -- we have already upgrade tests, so maybe it's sufficient to add a FK-join to this test topology to cover this case?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java (hunk 11 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java:30]: Good catch! Done in f7f9181ec3312b21a3534919819ccd02dfdcbc74.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java (hunk 12 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapper.java]: https://github.com/apache/kafka/pull/11945/commits/f7f9181ec3312b21a3534919819ccd02dfdcbc74#diff-f9c7a23b1e3f38e1a7c00d05f8d5fa725df7a8147a0d44818d3b63cbf7b0ad23R28

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java (hunk 13 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionWrapperSerde.java]: Fixed in https://github.com/apache/kafka/pull/11945/commits/1361497b55b2a34b094582f3220d54bb8e3e6c8a.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java (hunk 10 above)
Comment (Gerrrr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/foreignkeyjoin/SubscriptionResponseWrapper.java]: Fixed in https://github.com/apache/kafka/pull/11945/commits/f7f9181ec3312b21a3534919819ccd02dfdcbc74#diff-4a79c25b7d01993f00f2ffa554e15f18a2c4f6e1e0f5c18dc33ad24015261f22R24-R29

**Comment by mjsax** [comment]
Merged to `trunk`. Thanks for the fix!
